test/cloudtest/test_k8s_node_recovery.py

```diff
@@ -0,0 +1,303 @@
+        stalled_index,
+        must_exceed_reached_index=True,
+        timeout_in_sec=330,
+    )
+
+    recovered_index = get_current_counter_index(mz, cluster_setup)
+
+    mz.revive_suspended_k8s_node(suspended_node_name)
+
+    # expect no issues after node recovery
+    validate_state(
+        mz,
+        cluster_setup,
+        recovered_index + REQUIRED_PROGRESS,
+        must_exceed_reached_index=True,
+        timeout_in_sec=10,
+    )
+
+
+@pytest.mark.long
+@unittest.skip("Requires #20299 to work")
+def test_unreplicated_compute_cluster_on_failing_node(
+    mz: MaterializeApplication,
+) -> None:
+    """
+    An unreplicated compute cluster is on the failed node. Queries of indexes on the compute cluster should fail, but
+    resolve within a minute or two.
+    """
+    cluster_setup = ClusterSetup()
+    populate(mz, cluster_setup)
+    reached_index = get_current_counter_index(mz, cluster_setup)
+
+    suspended_node_name = suspend_first_node_of_pod(mz, cluster_setup.compute_name)
+
+    print("Awaiting node recovery...")
+    # expect (live) data to be delivered after at most after two minutes in production (or 300s in k8s)
+    validate_state(
+        mz,
+        cluster_setup,
+        reached_index + REQUIRED_PROGRESS,
+        must_exceed_reached_index=True,
+        timeout_in_sec=330,
+    )
+
+    recovered_index = get_current_counter_index(mz, cluster_setup)
+
+    mz.revive_suspended_k8s_node(suspended_node_name)
+
+    # expect no issues after node recovery
+    validate_state(
+        mz,
+        cluster_setup,
+        recovered_index + REQUIRED_PROGRESS,
+        must_exceed_reached_index=True,
+        timeout_in_sec=10,
+    )
+
+
+@pytest.mark.long
+def test_replicated_compute_cluster_on_failing_node(mz: MaterializeApplication) -> None:
+    """
+    A replicated compute cluster is on the failed node. Queries of indexes on the compute cluster should experience no
+    disruption in latency, thanks to the second replica.
+    """
+    cluster_setup = ClusterSetup(compute_replication_factor=2)
+    populate(mz, cluster_setup)
+    reached_index = get_current_counter_index(mz, cluster_setup)
+
+    suspended_node_name = suspend_first_node_of_pod(mz, cluster_setup.compute_name)
+
+    print("Awaiting node recovery...")
+    # expect (live) data to be delivered without disruption in latency
+    validate_state(
+        mz,
+        cluster_setup,
+        reached_index + REQUIRED_PROGRESS,
+        must_exceed_reached_index=True,
+        timeout_in_sec=30,
+    )
+
+    reached_index = get_current_counter_index(mz, cluster_setup)
+
+    mz.revive_suspended_k8s_node(suspended_node_name)
+
+    # expect no issues after node recovery
+    validate_state(
+        mz,
+        cluster_setup,
+        reached_index + REQUIRED_PROGRESS,
+        must_exceed_reached_index=True,
+        timeout_in_sec=10,
+    )
+
+
+@pytest.mark.long
+def test_envd_on_failing_node(mz: MaterializeApplication) -> None:
```

Comment:
At the end, I could not make this work. suspending a K8s node does not cause the environmentd pod to be rescheduled on another node. You need to also delete it, at which point it may be rescheduled on **exactly the same node** that is now NotReady ...

The only scenario I could make work is this one:

```
-def test_envd_on_failing_node(mz: MaterializeApplication) -> None:
+def test_envd_on_cordoned_node(mz: MaterializeApplication) -> None:
     """
-    environmentd is on the failed node. All connections should fail, but resolve within a minute or two.
+    environmentd is on a node that is being drained. All connections should fail, but resolve within a minute or two.
     """
     cluster_setup = ClusterSetup()
     populate(mz, cluster_setup)
@@ -255,7 +259,8 @@ def test_envd_on_failing_node(mz: MaterializeApplication) -> None:
     envd_node_name = mz.get_k8s_value(
         "app=environmentd", "{.items[*].spec.nodeName}", remove_quotes=True
     )
-    mz.suspend_k8s_node(envd_node_name)
+    mz.kubectl("cordon", envd_node_name)
+    mz.kubectl("delete", "--force", "pod", "environmentd-0")
 
     print("Expecting connection timeout...")
     # all connections / queries should fail initially
@@ -276,6 +281,8 @@ def test_envd_on_failing_node(mz: MaterializeApplication) -> None:
         pass
     print("Survived connection timeout.")
 
+    mz.kubectl("uncordon", envd_node_name)
+
     print("Awaiting node recovery...")
     # expect (live) data to be delivered after at most after two minutes in production (or 300s in k8s)
     validate_state(
@@ -285,16 +292,3 @@ def test_envd_on_failing_node(mz: MaterializeApplication) -> None:
         must_exceed_reached_index=True,
         timeout_in_sec=330,
     )
-
-    recovered_index = get_current_counter_index(mz, cluster_setup)
-
-    mz.revive_suspended_k8s_node(envd_node_name)
-
-    # expect no issues after node recovery
-    validate_state(
-        mz,
-        cluster_setup,
-        recovered_index + REQUIRED_PROGRESS,
-        must_exceed_reached_index=True,
-        timeout_in_sec=10,
-    )
```

This has the added benefit that only one K8s node is needed for environmentd.

Reply:
Discussed on Slack, but this should work if you set the `terminationGracePeriod: 0` setting on the `EnvironmentdStatefulSet` here:

https://github.com/MaterializeInc/materialize/blob/bcd56263ec81a12ad9f4bafcbe49043fbb0c148d/misc/python/materialize/cloudtest/k8s/environmentd.py#L244-L255

(This is what we do in production as of recently.)

Reply:
✅ Added. (To my understanding, it needs to be added to the _pod spec_.)